File: 2020/13/main_test.go
package main

import (
	"testing"
)

func TestBusShouldHaveTimetable(t *testing.T) {
	start := 939
	tests := []Bus{7, 13, 19, 31, 59}
	expected := []int{945, 949, 950, 961, 944}

	for i, test := range tests {
		result := test.NextDeparture(start)
		if result != expected[i] {
			t.Errorf("Bus %d: got %d, expected %d", test, result, expected[i])
		}
	}
}

func TestTask1ShouldFindResult(t *testing.T) {
	start := 939
	input := []Bus{7, 13, -1, -1, 59, -1, 31, 19}
	expected := 295

	result := Task1(start, input)
	if result != expected {
		t.Errorf("got %d, expected %d", result, expected)
	}
}

func TestTask2ShouldFindResult(t *testing.T) {
	tests := [][]Bus{
		{7, 13, -1, -1, 59, -1, 31, 19},
		{17, -1, 13, 19},
		{67, 7, 59, 61},
		{67, -1, 7, 59, 61},
		{67, 7, -1, 59, 61},
		{1789, 37, 47, 1889},
	}
	expected := []int{1068781, 3417, 754018, 779210, 1261476, 1202161486}

	for i, test := range tests {
		result := Task2(test)
		if result != expected[i] {
			t.Errorf("%d: got %d, expected %d", i, result, expected)
		}
	}
}

File: 2021/12/src/main_test.zig
const std = @import("std");

const app = @import("./main.zig");

test "expect task 1 test 1 to result in 10" {
    std.testing.log_level = .debug;

    const raw_input: []const u8 =
        \\start-A
        \\start-b
        \\A-c
        \\A-b
        \\b-d
        \\A-end
        \\b-end
        \\
    ;
    var stream = std.io.fixedBufferStream(raw_input);
    var input = try app.readInput(std.testing.allocator, stream.reader());
    defer input.deinit();
    const expected: u32 = 10;

    try std.testing.expect((try app.task1(std.testing.allocator, &input)) == expected);
}

test "expect task 1 test 2 to result in 19" {
    std.testing.log_level = .debug;

    const raw_input: []const u8 =
        \\dc-end
        \\HN-start
        \\start-kj
        \\dc-start
        \\dc-HN
        \\LN-dc
        \\HN-end
        \\kj-sa
        \\kj-HN
        \\kj-dc
        \\
    ;
    var stream = std.io.fixedBufferStream(raw_input);
    var input = try app.readInput(std.testing.allocator, stream.reader());
    defer input.deinit();
    const expected: u32 = 19;

    try std.testing.expect((try app.task1(std.testing.allocator, &input)) == expected);
}

test "expect task 1 test 3 to result in 226" {
    std.testing.log_level = .debug;

    const raw_input: []const u8 =
        \\fs-end
        \\he-DX
        \\fs-he
        \\start-DX
        \\pj-DX
        \\end-zg
        \\zg-sl
        \\zg-pj
        \\pj-he
        \\RW-he
        \\fs-DX
        \\pj-RW
        \\zg-RW
        \\start-pj
        \\he-WI
        \\zg-he
        \\pj-fs
        \\start-RW
        \\
    ;
    var stream = std.io.fixedBufferStream(raw_input);
    var input = try app.readInput(std.testing.allocator, stream.reader());
    defer input.deinit();
    const expected: u32 = 226;

    try std.testing.expect((try app.task1(std.testing.allocator, &input)) == expected);
}

test "expect task 2 test 1 to result in 36" {
    std.testing.log_level = .debug;

    const raw_input: []const u8 =
        \\start-A
        \\start-b
        \\A-c
        \\A-b
        \\b-d
        \\A-end
        \\b-end
        \\
    ;
    var stream = std.io.fixedBufferStream(raw_input);
    var input = try app.readInput(std.testing.allocator, stream.reader());
    defer input.deinit();
    const expected: u32 = 36;

    try std.testing.expect((try app.task2(std.testing.allocator, &input)) == expected);
}

test "expect task 2 test 2 to result in 103" {
    std.testing.log_level = .debug;

    const raw_input: []const u8 =
        \\dc-end
        \\HN-start
        \\start-kj
        \\dc-start
        \\dc-HN
        \\LN-dc
        \\HN-end
        \\kj-sa
        \\kj-HN
        \\kj-dc
        \\
    ;
    var stream = std.io.fixedBufferStream(raw_input);
    var input = try app.readInput(std.testing.allocator, stream.reader());
    defer input.deinit();
    const expected: u32 = 103;

    try std.testing.expect((try app.task2(std.testing.allocator, &input)) == expected);
}

test "expect task 2 test 3 to result in 3509" {
    std.testing.log_level = .debug;

    const raw_input: []const u8 =
        \\fs-end
        \\he-DX
        \\fs-he
        \\start-DX
        \\pj-DX
        \\end-zg
        \\zg-sl
        \\zg-pj
        \\pj-he
        \\RW-he
        \\fs-DX
        \\pj-RW
        \\zg-RW
        \\start-pj
        \\he-WI
        \\zg-he
        \\pj-fs
        \\start-RW
        \\
    ;
    var stream = std.io.fixedBufferStream(raw_input);
    var input = try app.readInput(std.testing.allocator, stream.reader());
    defer input.deinit();
    const expected: u32 = 3509;

    try std.testing.expect((try app.task2(std.testing.allocator, &input)) == expected);
}
